kong/db/strategies/cassandra/connector.lua

```diff
@@ -14,19 +14,105 @@ function CassandraConnector.new(kong_config)
-
-      else
-        log.debug("resolved Cassandra contact point '%s' to: %s", cp, ip)
-        resolved_contact_points[i] = ip
+
+    -- TODO: this is an ugly hack to force lua sockets on a third party library
+
+    local pok, perr = pcall(function()
+      local tcp_old = ngx.socket.tcp
+      local udp_old = ngx.socket.udp
+
+      local dns_no_sync_old = kong_config.dns_no_sync
+
+      package.loaded["socket"] = nil
+      package.loaded["kong.tools.dns"] = nil
+      package.loaded["resty.dns.client"] = nil
+      package.loaded["resty.dns.resolver"] = nil
+
+      ngx.socket.tcp = function(...)
+        local tcp = require("socket").tcp(...)
+        return setmetatable({}, {
+          __newindex = function(_, k, v)
+            tcp[k] = v
+          end,
+          __index = function(_, k)
+            if type(tcp[k]) == "function" then
+              return function(_, ...)
+                if k == "send" then
+                  local value = select(1, ...)
+                  if type(value) == "table" then
+                    return tcp.send(tcp, table.concat(value))
+                  end
+
+                  return tcp.send(tcp, ...)
+                end
+
+                return tcp[k](tcp, ...)
+              end
+            end
+
+            return tcp[k]
+          end
+        })
+      end
+
+      ngx.socket.udp = function(...)
+        local udp = require("socket").udp(...)
+        return setmetatable({}, {
+          __newindex = function(_, k, v)
+            udp[k] = v
+          end,
+          __index = function(_, k)
+            if type(udp[k]) == "function" then
+              return function(_, ...)
+                if k == "send" then
+                  local value = select(1, ...)
+                  if type(value) == "table" then
+                    return udp.send(udp, table.concat(value))
+                  end
+
+                  return udp.send(udp, ...)
+                end
+
+                return udp[k](udp, ...)
+              end
+            end
+
+            return udp[k]
+          end
+        })
+      end
+
+      local dns_tools = require "kong.tools.dns"
+
+      kong_config.dns_no_sync = true
+
+      local dns = dns_tools(kong_config)
+
+      for i, cp in ipairs(kong_config.cassandra_contact_points) do
+        local ip, err = dns.toip(cp)
+        if not ip then
+          log.error("could not resolve Cassandra contact point '%s': %s",
+                    cp, err)
+
+        else
+          log.debug("resolved Cassandra contact point '%s' to: %s", cp, ip)
+          resolved_contact_points[i] = ip
+        end
       end
+
+      kong_config.dns_no_sync = dns_no_sync_old
+
+      package.loaded["resty.dns.resolver"] = nil
+      package.loaded["resty.dns.client"] = nil
+      package.loaded["kong.tools.dns"] = nil
+      package.loaded["socket"] = nil
+
+      ngx.socket.udp = udp_old
+      ngx.socket.tcp = tcp_old
+    end)
+
+    if not pok then
+      return nil, perr
```

Comment:
We should return a more descriptive error here, maybe something like:

```lua
return nil, "error while patching resty.dns.resolver and resolving contact points: " .. perr
```

I actually even wonder what the rationale is behind protecting this block? It hides the stacktraces and a returned error produces the same result as a runtime one: Kong cannot start because it has no DAO. What is the rationale/how does it make the booting sequence more resilient?

Reply:
removed pcall.